### ISSUE TITLE
wire: test flakes on Travis with "failed to cache compiled Go files"

### ISSUE THREAD

**Issue by zombiezen** [google/go-cloud#669]
### Describe the bug

The Wire tests fail intermittently on Travis with an error like:

```
        wire_test.go:96: go [list -e -json -compiled -test=false -export=false -deps=true -tags=wireinject -- example.com/foo]: exit status 1: go: finding github.com/google/go-cloud v0.2.0
            go build runtime/internal/atomic: failed to cache compiled Go files
```

### To Reproduce

This is an intermittent failure, but I've only seen it on Travis. Uncertain what the triggers are.

## Expected behavior

Wire tests should pass.

/cc @vangent, who I believe hit this failure before, but I can't find evidence of it.

**Comment by vangent** [comment]
https://travis-ci.com/google/go-cloud/jobs/157991519

shows this.

**Comment by zombiezen** [comment]
Yeah, worth noting that the test it fails on is not fixed, and the package that fails is not fixed. Only the error message.

Perhaps there's contention running `go list` a lot concurrently?

**Comment by zombiezen** [comment]
@alandonovan @ianthehat, any ideas on what's going on here? Seems like the conditions in here are triggering regularly:

https://github.com/golang/go/blob/0a72e8eefbdaa62bf782618d50851e4740408cba/src/cmd/go/internal/work/exec.go#L599-L604

Are we holding something wrong or is this an issue with `go/packages`?

**Comment by ianthehat** [comment]
Not really sure, but I know that there are known issues with running go list in module mode concurrently, that do tend to cause errors like this, is that a possibility?
This means you cant use go/packages.Load concurrently right now, from either a single process or multiple processes

**Comment by ianthehat** [comment]
It's the go command (you end up with a corrupted module cache), and I know it's what Bryan has been focusing on for a while now, I think he is supposed to be done with making it work, so hopefully it will all be in 1.12, but you should check with him to be sure.
I don't know what the tracking issue is though, sorry.

**Comment by zombiezen** [comment]
@bcmills, is there a Go bug tracking the corrupted module cache? I am slightly puzzled why this would occur, as each test runs in a separate, isolated `GOPATH`.

**Comment by bcmills** [comment]
The known source of module-cache corruption is https://github.com/golang/go/issues/26794.

The `go: finding` message relates to the module cache, but the `failed to cache compiled Go files` message is the build cache.

**Comment by zombiezen** [comment]
Thanks everyone!

My plan is to remove the parallelism in the test for the time being, see if that addresses the issue we're seeing, and if so, keep the workaround in place until the upstream issue is fixed. Expect a patch soon.

**Comment by zombiezen** [comment]
Oh, now I understand the nuance. I'll see if I can give you some reproduction steps. Still carrying forward with the workaround, though.

**Comment by zombiezen** [comment]
I'm not sure how to reproduce this yet. I've been trying variations of this command on my local box on a checkout of 0fd7870d6b88942a01c54e70afa3fe51cc020dc6 (before my workaround submitted):

```go
for i in $(seq 20); do rm -rf /tmp/gocache && echo "Iteration $i" && GOCACHE=/tmp/gocache go test -run=TestWire/'StructPointer|ValueConversion' ./wire/internal/wire; done
```

Nothing so far. The set of conditions leading to this are: empty cache, same GOCACHE (because it's unset), same GOROOT, lots of different isolated GOPATHs, many packages with the same import paths in those different GOPATHs, and all running concurrently on a box with limited resources.
